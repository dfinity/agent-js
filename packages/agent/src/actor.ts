import { Buffer } from 'buffer/';
import {
  Agent,
  getDefaultAgent,
  QueryResponseRejected,
  QueryResponseStatus,
  ReplicaRejectCode,
  SubmitResponse,
} from './agent';
import { AgentError } from './errors';
import { IDL } from '@dfinity/candid';
import { pollForResponse, PollStrategyFactory, strategy } from './polling';
import { Principal } from '@dfinity/principal';
import { RequestId } from './request_id';
import { toHex } from './utils/buffer';
import { CreateCertificateOptions } from './certificate';
import managementCanisterIdl from './canisters/management_idl';
import _SERVICE from './canisters/management_service';

export class ActorCallError extends AgentError {
  constructor(
    public readonly canisterId: Principal,
    public readonly methodName: string,
    public readonly type: 'query' | 'update',
    public readonly props: Record<string, string>,
  ) {
    super(
      [
        `Call failed:`,
        `  Canister: ${canisterId.toText()}`,
        `  Method: ${methodName} (${type})`,
        ...Object.getOwnPropertyNames(props).map(n => `  "${n}": ${JSON.stringify(props[n])}`),
      ].join('\n'),
    );
  }
}

export class QueryCallRejectedError extends ActorCallError {
  constructor(
    canisterId: Principal,
    methodName: string,
    public readonly result: QueryResponseRejected,
  ) {
    super(canisterId, methodName, 'query', {
      Status: result.status,
      Code: ReplicaRejectCode[result.reject_code] ?? `Unknown Code "${result.reject_code}"`,
      Message: result.reject_message,
    });
  }
}

export class UpdateCallRejectedError extends ActorCallError {
  constructor(
    canisterId: Principal,
    methodName: string,
    public readonly requestId: RequestId,
    public readonly response: SubmitResponse['response'],
  ) {
    super(canisterId, methodName, 'update', {
      'Request ID': toHex(requestId),
      'HTTP status code': response.status.toString(),
      'HTTP status text': response.statusText,
    });
  }
}

/**
 * Configuration to make calls to the Replica.
 */
export interface CallConfig {
  /**
   * An agent to use in this call, otherwise the actor or call will try to discover the
   * agent to use.
   */
  agent?: Agent;

  /**
   * A polling strategy factory that dictates how much and often we should poll the
   * read_state endpoint to get the result of an update call.
   */
  pollingStrategyFactory?: PollStrategyFactory;

  /**
   * The canister ID of this Actor.
   */
  canisterId?: string | Principal;

  /**
   * The effective canister ID. This should almost always be ignored.
   */
  effectiveCanisterId?: Principal;
}

/**
 * Configuration that can be passed to customize the Actor behaviour.
 */
export interface ActorConfig extends CallConfig {
  /**
   * The Canister ID of this Actor. This is required for an Actor.
   */
  canisterId: string | Principal;

  /**
   * An override function for update calls' CallConfig. This will be called on every calls.
   */
  callTransform?(
    methodName: string,
    args: unknown[],
    callConfig: CallConfig,
  ): Partial<CallConfig> | void;

  /**
   * An override function for query calls' CallConfig. This will be called on every query.
   */
  queryTransform?(
    methodName: string,
    args: unknown[],
    callConfig: CallConfig,
  ): Partial<CallConfig> | void;

  /**
   * Polyfill for BLS Certificate verification in case wasm is not supported
   */
  blsVerify?: CreateCertificateOptions['blsVerify'];
}

// TODO: move this to proper typing when Candid support TypeScript.
/**
 * A subclass of an actor. Actor class itself is meant to be a based class.
 */
export type ActorSubclass<T = Record<string, ActorMethod>> = Actor & T;

/**
 * An actor method type, defined for each methods of the actor service.
 */
export interface ActorMethod<Args extends unknown[] = unknown[], Ret extends unknown = unknown> {
  (...args: Args): Promise<Ret>;
  withOptions(options: CallConfig): (...args: Args) => Promise<Ret>;
}

/**
 * The mode used when installing a canister.
 */
export enum CanisterInstallMode {
  Install = 'install',
  Reinstall = 'reinstall',
  Upgrade = 'upgrade',
}

/**
 * Internal metadata for actors. It's an enhanced version of ActorConfig with
 * some fields marked as required (as they are defaulted) and canisterId as
 * a Principal type.
 */
interface ActorMetadata {
  service: IDL.ServiceClass;
  agent?: Agent;
  config: ActorConfig;
}

const metadataSymbol = Symbol.for('ic-agent-metadata');

/**
 * An actor base class. An actor is an object containing only functions that will
 * return a promise. These functions are derived from the IDL definition.
 */
export class Actor {
  /**
   * Get the Agent class this Actor would call, or undefined if the Actor would use
   * the default agent (global.ic.agent).
   * @param actor The actor to get the agent of.
   */
  public static agentOf(actor: Actor): Agent | undefined {
    return actor[metadataSymbol].config.agent;
  }

  /**
   * Get the interface of an actor, in the form of an instance of a Service.
   * @param actor The actor to get the interface of.
   */
  public static interfaceOf(actor: Actor): IDL.ServiceClass {
    return actor[metadataSymbol].service;
  }

  public static canisterIdOf(actor: Actor): Principal {
    return Principal.from(actor[metadataSymbol].config.canisterId);
  }

  public static async install(
    fields: {
      module: ArrayBuffer;
      mode?: CanisterInstallMode;
      arg?: ArrayBuffer;
    },
    config: ActorConfig,
  ): Promise<void> {
    const mode = fields.mode === undefined ? CanisterInstallMode.Install : fields.mode;
    // Need to transform the arg into a number array.
    const arg = fields.arg ? [...new Uint8Array(fields.arg)] : [];
    // Same for module.
    const wasmModule = [...new Uint8Array(fields.module)];
    const canisterId =
      typeof config.canisterId === 'string'
        ? Principal.fromText(config.canisterId)
        : config.canisterId;

    await getManagementCanister(config).install_code({
      mode: { [mode]: null } as any,
      arg,
      wasm_module: wasmModule,
      canister_id: canisterId,
    });
  }

  public static async createCanister(config?: CallConfig): Promise<Principal> {
    const { canister_id: canisterId } = await getManagementCanister(
      config || {},
    ).provisional_create_canister_with_cycles({ amount: [], settings: [] });

    return canisterId;
  }

  public static async createAndInstallCanister(
    interfaceFactory: IDL.InterfaceFactory,
    fields: {
      module: ArrayBuffer;
      arg?: ArrayBuffer;
    },
    config?: CallConfig,
  ): Promise<ActorSubclass> {
    const canisterId = await this.createCanister(config);
    await this.install(
      {
        ...fields,
      },
      { ...config, canisterId },
    );

    return this.createActor(interfaceFactory, { ...config, canisterId });
  }

  public static createActorClass(interfaceFactory: IDL.InterfaceFactory): ActorConstructor {
    const service = interfaceFactory({ IDL });

    class CanisterActor extends Actor {
      [x: string]: ActorMethod;

      constructor(config: ActorConfig) {
        const canisterId =
          typeof config.canisterId === 'string'
            ? Principal.fromText(config.canisterId)
            : config.canisterId;

        super({
          config: {
            ...DEFAULT_ACTOR_CONFIG,
            ...config,
            canisterId,
          },
          service,
        });

        for (const [methodName, func] of service._fields) {
          this[methodName] = _createActorMethod(this, methodName, func, config.blsVerify);
        }
      }
    }

    return CanisterActor;
  }

  public static createActor<T = Record<string, ActorMethod>>(
    interfaceFactory: IDL.InterfaceFactory,
    configuration: ActorConfig,
  ): ActorSubclass<T> {
    return new (this.createActorClass(interfaceFactory))(
      configuration,
    ) as unknown as ActorSubclass<T>;
  }

  private [metadataSymbol]: ActorMetadata;

  protected constructor(metadata: ActorMetadata) {
    this[metadataSymbol] = Object.freeze(metadata);
  }
}

// IDL functions can have multiple return values, so decoding always
// produces an array. Ensure that functions with single or zero return
// values behave as expected.
function decodeReturnValue(types: IDL.Type[], msg: ArrayBuffer) {
  const returnValues = IDL.decode(types, Buffer.from(msg));
  switch (returnValues.length) {
    case 0:
      return undefined;
    case 1:
      return returnValues[0];
    default:
      return returnValues;
  }
}

const DEFAULT_ACTOR_CONFIG = {
  pollingStrategyFactory: strategy.defaultStrategy,
};

export type ActorConstructor = new (config: ActorConfig) => ActorSubclass;

function _createActorMethod(
  actor: Actor,
  methodName: string,
  func: IDL.FuncClass,
  blsVerify?: CreateCertificateOptions['blsVerify'],
): ActorMethod {
  let caller: (options: CallConfig, ...args: unknown[]) => Promise<unknown>;
  if (func.annotations.includes('query')) {
    caller = async (options, ...args) => {
      // First, if there's a config transformation, call it.
      options = {
        ...options,
        ...actor[metadataSymbol].config.queryTransform?.(methodName, args, {
          ...actor[metadataSymbol].config,
          ...options,
        }),
      };

      const agent = options.agent || actor[metadataSymbol].config.agent || getDefaultAgent();
      const cid = Principal.from(options.canisterId || actor[metadataSymbol].config.canisterId);
      const arg = IDL.encode(func.argTypes, args);

      const result = await agent.query(cid, { methodName, arg });

      switch (result.status) {
        case QueryResponseStatus.Rejected:
          throw new QueryCallRejectedError(cid, methodName, result);

        case QueryResponseStatus.Replied:
          return decodeReturnValue(func.retTypes, result.reply.arg);
      }
    };
  } else {
    caller = async (options, ...args) => {
      // First, if there's a config transformation, call it.
      options = {
        ...options,
        ...actor[metadataSymbol].config.callTransform?.(methodName, args, {
          ...actor[metadataSymbol].config,
          ...options,
        }),
      };

      const agent = options.agent || actor[metadataSymbol].config.agent || getDefaultAgent();
      const { canisterId, effectiveCanisterId, pollingStrategyFactory } = {
        ...DEFAULT_ACTOR_CONFIG,
        ...actor[metadataSymbol].config,
        ...options,
      };
      const cid = Principal.from(canisterId);
      const ecid = effectiveCanisterId !== undefined ? Principal.from(effectiveCanisterId) : cid;
      const arg = IDL.encode(func.argTypes, args);
      const { requestId, response } = await agent.call(cid, {
        methodName,
        arg,
        effectiveCanisterId: ecid,
      });

      if (!response.ok) {
        throw new UpdateCallRejectedError(cid, methodName, requestId, response);
      }

      const pollStrategy = pollingStrategyFactory();
      const responseBytes = await pollForResponse(agent, ecid, requestId, pollStrategy, blsVerify);

      if (responseBytes !== undefined) {
        return decodeReturnValue(func.retTypes, responseBytes);
      } else if (func.retTypes.length === 0) {
        return undefined;
      } else {
        throw new Error(`Call was returned undefined, but type [${func.retTypes.join(',')}].`);
      }
    };
  }

  const handler = (...args: unknown[]) => caller({}, ...args);
  handler.withOptions =
    (options: CallConfig) =>
    (...args: unknown[]) =>
      caller(options, ...args);
  return handler as ActorMethod;
}

export type ManagementCanisterRecord = _SERVICE;

/**
 * Create a management canister actor
 * @param config
 */
export function getManagementCanister(config: CallConfig): ActorSubclass<ManagementCanisterRecord> {
  function transform(_methodName: string, args: unknown[], _callConfig: CallConfig) {
    const first = args[0] as any;
    let effectiveCanisterId = Principal.fromHex('');
    if (first && typeof first === 'object' && first.canister_id) {
      effectiveCanisterId = Principal.from(first.canister_id as unknown);
    }
    return { effectiveCanisterId };
  }

  return Actor.createActor<ManagementCanisterRecord>(managementCanisterIdl, {
    ...config,
    canisterId: Principal.fromHex(''),
    ...{
      callTransform: transform,
      queryTransform: transform,
    },
  });
}
