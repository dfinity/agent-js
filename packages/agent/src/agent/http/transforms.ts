import { lebEncode } from '@dfinity/candid';
import * as cbor from 'simple-cbor';
import { Endpoint, HttpAgentRequest, HttpAgentRequestTransformFn, makeNonce, Nonce } from './types';
import { toHex } from '../../utils/buffer';

const NANOSECONDS_PER_MILLISECONDS = BigInt(1_000_000);

const REPLICA_PERMITTED_DRIFT_MILLISECONDS = BigInt(60 * 1000);

export class Expiry {
  private readonly _value: bigint;

  constructor(deltaInMSec: number) {
    // Use bigint because it can overflow the maximum number allowed in a double float.
    this._value =
      (BigInt(Date.now()) + BigInt(deltaInMSec) - REPLICA_PERMITTED_DRIFT_MILLISECONDS) *
      NANOSECONDS_PER_MILLISECONDS;
  }

  public toCBOR(): cbor.CborValue {
    // TODO: change this to take the minimum amount of space (it always takes 8 bytes now).
    return cbor.value.u64(this._value.toString(16), 16);
  }

  public toHash(): ArrayBuffer {
    return lebEncode(this._value);
  }
}

/**
 * Create a Nonce transform, which takes a function that returns a Buffer, and adds it
 * as the nonce to every call requests.
 * @param nonceFn A function that returns a buffer. By default uses a semi-random method.
 */
export function makeNonceTransform(nonceFn: () => Nonce = makeNonce): HttpAgentRequestTransformFn {
  return async (request: HttpAgentRequest) => {
    const nonce = nonceFn();
    // Nonce needs to be inserted into the header for all requests, to enable logs to be correlated with requests.
    const headers = request.request.headers ? new Headers(request.request.headers) : new Headers();
    // TODO: uncomment this when the http proxy supports it.
    // headers.set('X-IC-Request-ID', toHex(new Uint8Array(nonce)));
    request.request.headers = headers;

    // Nonce only needs to be inserted into the body for async calls, to prevent replay attacks.
    if (request.endpoint === Endpoint.Call) {
      request.body.nonce = nonceFn();
    }
  };
}

/**
 * Create a transform that adds a delay (by default 5 minutes) to the expiry.
 *
 * @param delayInMilliseconds The delay to add to the call time, in milliseconds.
 */
export function makeExpiryTransform(delayInMilliseconds: number): HttpAgentRequestTransformFn {
  return async (request: HttpAgentRequest) => {
    request.body.ingress_expiry = new Expiry(delayInMilliseconds);
  };
}
