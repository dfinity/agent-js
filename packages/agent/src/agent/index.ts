import { GlobalInternetComputer } from '../index';
import { Agent } from './api';

export * from './api';
export * from './http';
export * from './proxy';

declare const window: GlobalInternetComputer;
declare const global: GlobalInternetComputer;
declare const self: GlobalInternetComputer;

export function getDefaultAgent(): Agent {
  const agent =
    typeof window === 'undefined'
      ? typeof global === 'undefined'
        ? typeof self === 'undefined'
          ? undefined
          : self.ic.agent
        : global.ic.agent
      : window.ic.agent;

  if (!agent) {
    throw new Error('No Agent could be found.');
  }

  return agent;
}
