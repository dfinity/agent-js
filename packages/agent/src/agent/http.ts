import { Buffer } from 'buffer/';
import { ActorFactory } from '../actor';
import * as actor from '../actor';
import { AnonymousIdentity, Identity } from '../auth';
import * as cbor from '../cbor';
import { Expiry } from '../http_agent_transforms';
import {
  Endpoint,
  HttpAgentRequest,
  HttpAgentRequestTransformFn,
  HttpAgentSubmitRequest,
  QueryFields,
  QueryResponse,
  ReadRequest,
  ReadRequestType,
  ReadResponse,
  ReadStateFields,
  ReadStateResponse,
  SubmitRequest,
  SubmitRequestType,
  SubmitResponse,
} from '../http_agent_types';
import * as IDL from '../idl';
import { Principal } from '../principal';
import { requestIdOf } from '../request_id';
import { BinaryBlob, blobFromHex, JsonObject } from '../types';
import { Agent } from './api';

const API_VERSION = 'v1';

// Default delta for ingress expiry is 5 minutes.
const DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS = 5 * 60 * 1000;

// HttpAgent options that can be used at construction.
export interface HttpAgentOptions {
  // Another HttpAgent to inherit configuration (pipeline and fetch) of. This
  // is only used at construction.
  source?: HttpAgent;

  // A surrogate to the global fetch function. Useful for testing.
  fetch?: typeof fetch;

  // The host to use for the client. By default, uses the same host as
  // the current page.
  host?: string;

  // The principal used to send messages. This cannot be empty at the request
  // time (will throw).
  identity?: Identity | Promise<Identity>;

  credentials?: {
    name: string;
    password?: string;
  };
}

declare const window: Window & { fetch: typeof fetch };
declare const global: { fetch: typeof fetch };
declare const self: { fetch: typeof fetch };

function getDefaultFetch(): typeof fetch {
  const result =
    typeof window === 'undefined'
      ? typeof global === 'undefined'
        ? typeof self === 'undefined'
          ? undefined
          : self.fetch.bind(self)
        : global.fetch.bind(global)
      : window.fetch.bind(window);

  if (!result) {
    throw new Error('Could not find default `fetch` implementation.');
  }

  return result;
}

// A HTTP agent allows users to interact with a client of the internet computer
// using the available methods. It exposes an API that closely follows the
// public view of the internet computer, and is not intended to be exposed
// directly to the majority of users due to its low-level interface.
//
// There is a pipeline to apply transformations to the request before sending
// it to the client. This is to decouple signature, nonce generation and
// other computations so that this class can stay as simple as possible while
// allowing extensions.
export class HttpAgent implements Agent {
  private readonly _pipeline: HttpAgentRequestTransformFn[] = [];
  private readonly _identity: Promise<Identity>;
  private readonly _fetch: typeof fetch;
  private readonly _host: URL;
  private readonly _credentials: string | undefined;

  constructor(options: HttpAgentOptions = {}) {
    if (options.source) {
      this._pipeline = [...options.source._pipeline];
      this._identity = options.source._identity;
    }
    this._fetch = options.fetch || getDefaultFetch() || fetch.bind(global);
    if (options.host) {
      if (!options.host.match(/^[a-z]+:/) && typeof window !== 'undefined') {
        this._host = new URL(window.location.protocol + '//' + options.host);
      } else {
        this._host = new URL(options.host);
      }
    } else {
      const location = window?.location;
      if (!location) {
        throw new Error('Must specify a host to connect to.');
      }
      this._host = new URL(location + '');
    }
    if (options.credentials) {
      const { name, password } = options.credentials;
      this._credentials = `${name}${password ? ':' + password : ''}`;
    }
    this._identity = Promise.resolve(options.identity || new AnonymousIdentity());
  }

  public addTransform(fn: HttpAgentRequestTransformFn, priority = fn.priority || 0) {
    // Keep the pipeline sorted at all time, by priority.
    const i = this._pipeline.findIndex(x => (x.priority || 0) < priority);
    this._pipeline.splice(i >= 0 ? i : this._pipeline.length, 0, Object.assign(fn, { priority }));
  }

  public async getPrincipal(): Promise<Principal | null> {
    return this._identity ? (await this._identity).getPrincipal() : Principal.anonymous();
  }

  public async call(
    canisterId: Principal | string,
    fields: {
      methodName: string;
      arg: BinaryBlob;
    },
    identity?: Identity | Promise<Identity>,
  ): Promise<SubmitResponse> {
    const id = await (identity !== undefined ? identity : this._identity);
    const sender = id?.getPrincipal() || Principal.anonymous();

    return this.submit(
      {
        request_type: SubmitRequestType.Call,
        canister_id: typeof canisterId === 'string' ? Principal.fromText(canisterId) : canisterId,
        method_name: fields.methodName,
        arg: fields.arg,
        sender: sender.toBlob(),
        ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
      },
      id,
    );
  }

  public async install(
    canisterId: Principal | string,
    fields: {
      module: BinaryBlob;
      arg?: BinaryBlob;
    },
    identity?: Identity | Promise<Identity>,
  ): Promise<SubmitResponse> {
    const id = await (identity || this._identity);
    const sender = id?.getPrincipal() || Principal.anonymous();
    return this.submit(
      {
        request_type: SubmitRequestType.InstallCode,
        canister_id: typeof canisterId === 'string' ? Principal.fromText(canisterId) : canisterId,
        module: fields.module,
        arg: fields.arg || blobFromHex(''),
        sender: sender.toBlob(),
        ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
      },
      id,
    );
  }

  public async createCanister(identity?: Identity | Promise<Identity>): Promise<SubmitResponse> {
    const id = await (identity || this._identity);
    const sender = id?.getPrincipal() || Principal.anonymous();

    return this.submit(
      {
        request_type: SubmitRequestType.CreateCanister,
        sender: sender.toBlob(),
        ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
      },
      id,
    );
  }

  public async query(
    canisterId: Principal | string,
    fields: QueryFields,
    identity?: Identity | Promise<Identity>,
  ): Promise<QueryResponse> {
    const id = await (identity || this._identity);
    const sender = id?.getPrincipal() || Principal.anonymous();

    return this.read(
      {
        request_type: ReadRequestType.Query,
        canister_id: typeof canisterId === 'string' ? Principal.fromText(canisterId) : canisterId,
        method_name: fields.methodName,
        arg: fields.arg,
        sender: sender.toBlob(),
        ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
      },
      id,
    ) as Promise<QueryResponse>;
  }

  public async readState(
    fields: ReadStateFields,
    identity?: Identity | Promise<Identity>,
  ): Promise<ReadStateResponse> {
    const id = await (identity || this._identity);
    const sender = id?.getPrincipal() || Principal.anonymous();

    return this.read(
      {
        request_type: ReadRequestType.ReadState,
        paths: fields.paths,
        sender: sender.toBlob(),
        ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
      },
      id,
    ) as Promise<ReadStateResponse>;
  }

  public async status(): Promise<JsonObject> {
    const headers: Record<string, string> = this._credentials
      ? {
          Authorization: 'Basic ' + btoa(this._credentials),
        }
      : {};

    const response = await this._fetch(
      '' + new URL(`/api/${API_VERSION}/${Endpoint.Status}`, this._host),
      { headers },
    );

    if (!response.ok) {
      throw new Error(
        `Server returned an error:\n` +
          `  Code: ${response.status} (${response.statusText}\n)` +
          `  Body: ${await response.text()}\n`,
      );
    }

    const buffer = await response.arrayBuffer();
    return cbor.decode(new Uint8Array(buffer));
  }

  public makeActorFactory(actorInterfaceFactory: IDL.InterfaceFactory): ActorFactory {
    return actor.makeActorFactory(actorInterfaceFactory);
  }

  protected _transform(request: HttpAgentRequest): Promise<HttpAgentRequest> {
    let p = Promise.resolve(request);

    for (const fn of this._pipeline) {
      p = p.then(r => fn(r).then(r2 => r2 || r));
    }

    return p;
  }

  protected async submit(submit: SubmitRequest, identity: Identity): Promise<SubmitResponse> {
    let transformedRequest: any = (await this._transform({
      request: {
        body: null,
        method: 'POST',
        headers: {
          'Content-Type': 'application/cbor',
          ...(this._credentials ? { Authorization: 'Basic ' + btoa(this._credentials) } : {}),
        },
      },
      endpoint: Endpoint.Submit,
      body: submit,
    })) as HttpAgentSubmitRequest;

    // Apply transform for identity.
    transformedRequest = await identity.transformRequest(transformedRequest);

    const body = cbor.encode(transformedRequest.body);

    // Run both in parallel. The fetch is quite expensive, so we have plenty of time to
    // calculate the requestId locally.
    const [response, requestId] = await Promise.all([
      this._fetch('' + new URL(`/api/${API_VERSION}/${Endpoint.Submit}`, this._host), {
        ...transformedRequest.request,
        body,
      }),
      requestIdOf(submit),
    ]);

    if (!response.ok) {
      throw new Error(
        `Server returned an error:\n` +
          `  Code: ${response.status} (${response.statusText}\n)` +
          `  Body: ${await response.text()}\n`,
      );
    }

    return {
      requestId,
      response: {
        ok: response.ok,
        status: response.status,
        statusText: response.statusText,
      },
    };
  }

  protected async read(request: ReadRequest, identity: Identity): Promise<ReadResponse> {
    // TODO: remove this any. This can be a Signed or UnSigned request.
    let transformedRequest: any = await this._transform({
      request: {
        method: 'POST',
        headers: {
          'Content-Type': 'application/cbor',
          ...(this._credentials ? { Authorization: 'Basic ' + btoa(this._credentials) } : {}),
        },
      },
      endpoint: Endpoint.Read,
      body: request,
    });

    // Apply transform for identity.
    transformedRequest = await identity.transformRequest(transformedRequest);

    const body = cbor.encode(transformedRequest.body);

    const response = await this._fetch(
      '' + new URL(`/api/${API_VERSION}/${Endpoint.Read}`, this._host),
      {
        ...transformedRequest.request,
        body,
      },
    );

    if (!response.ok) {
      throw new Error(
        `Server returned an error:\n` +
          `  Code: ${response.status} (${response.statusText}\n)` +
          `  Body: ${await response.text()}\n`,
      );
    }
    return cbor.decode(Buffer.from(await response.arrayBuffer()));
  }
}
