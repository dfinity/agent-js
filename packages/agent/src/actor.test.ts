import { IDL } from '@dfinity/candid';
import { Principal } from '@dfinity/principal';
import { Actor } from './actor';
import { HttpAgent, Nonce } from './agent';
import { Expiry, makeNonceTransform } from './agent/http/transforms';
import { CallRequest, SubmitRequestType, UnSigned } from './agent/http/types';
import * as cbor from './cbor';
import { requestIdOf } from './request_id';

const originalDateNowFn = global.Date.now;
beforeEach(() => {
  global.Date.now = jest.fn(() => new Date(1000000).getTime());
});
afterEach(() => {
  global.Date.now = originalDateNowFn;
});

describe('makeActor', () => {
  // TODO: update tests to be compatible with changes to Certificate
  it.skip('should encode calls', async () => {
    const actorInterface = () => {
      return IDL.Service({
        greet: IDL.Func([IDL.Text], [IDL.Text]),
      });
    };

    const expectedReplyArg = IDL.encode([IDL.Text], ['Hello, World!']);

    const mockFetch: jest.Mock = jest
      .fn()
      .mockImplementationOnce((/*resource, init*/) => {
        return Promise.resolve(
          new Response(null, {
            status: 202,
          }),
        );
      })
      .mockImplementationOnce((resource, init) => {
        const body = cbor.encode({ status: 'received' });
        return Promise.resolve(
          new Response(body, {
            status: 200,
          }),
        );
      })
      .mockImplementationOnce((resource, init) => {
        const body = cbor.encode({ status: 'processing' });
        return Promise.resolve(
          new Response(body, {
            status: 200,
          }),
        );
      })
      .mockImplementationOnce((resource, init) => {
        const body = cbor.encode({
          status: 'replied',
          reply: {
            arg: expectedReplyArg,
          },
        });
        return Promise.resolve(
          new Response(body, {
            status: 200,
          }),
        );
      });

    const methodName = 'greet';
    const argValue = 'Name';

    const arg = IDL.encode([IDL.Text], [argValue]);

    const canisterId = Principal.fromText('2chl6-4hpzw-vqaaa-aaaaa-c');
    const principal = await Principal.anonymous();
    const sender = principal.toUint8Array();

    const nonces = [
      new Uint8Array([0, 1, 2, 3, 4, 5, 6, 7]) as Nonce,
      new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8]) as Nonce,
      new Uint8Array([2, 3, 4, 5, 6, 7, 8, 9]) as Nonce,
      new Uint8Array([3, 4, 5, 6, 7, 8, 9, 0]) as Nonce,
      new Uint8Array([4, 5, 6, 7, 8, 9, 0, 1]) as Nonce,
    ];

    const expectedCallRequest = {
      content: {
        request_type: SubmitRequestType.Call,
        canister_id: canisterId,
        method_name: methodName,
        arg,
        nonce: nonces[0],
        sender,
        ingress_expiry: new Expiry(300000),
      },
    } as UnSigned<CallRequest>;

    const expectedCallRequestId = await requestIdOf(expectedCallRequest.content);

    let nonceCount = 0;

    const httpAgent = new HttpAgent({ fetch: mockFetch, disableNonce: true });
    httpAgent.addTransform(makeNonceTransform(() => nonces[nonceCount++]));

    const actor = Actor.createActor(actorInterface, { canisterId, agent: httpAgent });
    const reply = await actor.greet(argValue);

    expect(reply).toEqual(IDL.decode([IDL.Text], expectedReplyArg)[0]);

    const { calls } = mockFetch.mock;

    expect(calls.length).toBe(5);
    expect(calls[0]).toEqual([
      `http://localhost/api/v2/canister/${canisterId.toText()}/call`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/cbor',
        },
        body: cbor.encode(expectedCallRequest),
      },
    ]);

    expect(calls[1]).toEqual([
      `http://localhost/api/v2/canister/${canisterId.toText()}/read_state`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/cbor',
        },
        body: cbor.encode({
          content: {
            request_type: 'request_status',
            request_id: expectedCallRequestId,
            ingress_expiry: new Expiry(300000),
          },
        }),
      },
    ]);

    expect(calls[2][0]).toBe('http://localhost/api/v1/read');
    expect(calls[2][1]).toEqual({
      method: 'POST',
      headers: {
        'Content-Type': 'application/cbor',
      },
      body: cbor.encode({
        content: {
          request_type: 'request_status',
          request_id: expectedCallRequestId,
          ingress_expiry: new Expiry(300000),
        },
      }),
    });

    expect(calls[3][0]).toBe('http://localhost/api/v1/read');
    expect(calls[3][1]).toEqual({
      method: 'POST',
      headers: {
        'Content-Type': 'application/cbor',
      },
      body: cbor.encode({
        content: {
          request_type: 'request_status',
          request_id: expectedCallRequestId,
          ingress_expiry: new Expiry(300000),
        },
      }),
    });

    expect(calls[4][0]).toBe('http://localhost/api/v1/read');
    expect(calls[4][1]).toEqual({
      method: 'POST',
      headers: {
        'Content-Type': 'application/cbor',
      },
      body: cbor.encode({
        content: {
          request_type: 'request_status',
          request_id: expectedCallRequestId,
          ingress_expiry: new Expiry(300000),
        },
      }),
    });
  });
  it('should allow its agent to be invalidated', async () => {
    const mockFetch = jest.fn();
    const actorInterface = () => {
      return IDL.Service({
        greet: IDL.Func([IDL.Text], [IDL.Text]),
      });
    };
    const httpAgent = new HttpAgent({ fetch: mockFetch, host: 'http://localhost' });
    const canisterId = Principal.fromText('2chl6-4hpzw-vqaaa-aaaaa-c');
    const actor = Actor.createActor(actorInterface, { canisterId, agent: httpAgent });

    Actor.agentOf(actor).invalidateIdentity();

    try {
      await actor.greet('test');
    } catch (error) {
      expect(error.message).toBe(
        "This identity has expired due this application's security policy. Please refresh your authentication.",
      );
    }
  });
});

// TODO: tests for rejected, unknown time out
