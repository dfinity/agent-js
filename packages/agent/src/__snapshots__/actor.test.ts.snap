// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeActor should enrich actor interface with httpDetails 2`] = `
{
  "body": null,
  "headers": [],
  "ok": true,
  "requestDetails": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      4,
      116,
      101,
      115,
      116,
    ],
    "canister_id": {
      "__principal__": "2chl6-4hpzw-vqaaa-aaaaa-c",
    },
    "ingress_expiry": Expiry {
      "_value": 1200000000000n,
    },
    "method_name": "greet_update",
    "nonce": Uint8Array [],
    "request_type": "call",
    "sender": {
      "__principal__": "2vxsx-fae",
    },
  },
  "status": 202,
  "statusText": "accepted",
}
`;
